dspace-swordv2/src/main/java/org/dspace/sword2/VersionManager.java

```diff
@@ -10,75 +10,90 @@
 import org.dspace.authorize.AuthorizeException;
 import org.dspace.content.Bitstream;
 import org.dspace.content.Bundle;
+import org.dspace.content.BundleBitstream;
 import org.dspace.content.Item;
+import org.dspace.content.factory.ContentServiceFactory;
+import org.dspace.content.service.BitstreamService;
+import org.dspace.content.service.BundleService;
+import org.dspace.content.service.ItemService;
 import org.dspace.core.ConfigurationManager;
+import org.dspace.core.Context;
 
 import java.io.IOException;
 import java.sql.SQLException;
 import java.text.SimpleDateFormat;
 import java.util.Date;
+import java.util.List;
 
 public class VersionManager
 {
-	public void removeBundle(Item item, String name)
+	protected ItemService itemService = ContentServiceFactory.getInstance().getItemService();
+	protected BundleService bundleService = ContentServiceFactory.getInstance().getBundleService();
+	protected BitstreamService bitstreamService = ContentServiceFactory.getInstance().getBitstreamService();
+
+	public void removeBundle(Context context, Item item, String name)
 			throws SQLException, AuthorizeException, IOException
 	{
 		boolean keep = ConfigurationManager.getBooleanProperty("swordv2-server", "versions.keep");
-		Bundle[] bundles = item.getBundles(name);
+		List<Bundle> bundles = item.getBundles();
 		for (Bundle b : bundles)
 		{
-			this.removeBundle(item, b, keep);
+			if (name.equals(b.getName()))
+			{
+				this.removeBundle(context, item, b, keep);
+			}
 		}
 	}
 
-	public void removeBundle(Item item, Bundle source)
+	public void removeBundle(Context context, Item item, Bundle source)
 			throws SQLException, AuthorizeException, IOException
 	{
 		boolean keep = ConfigurationManager.getBooleanProperty("swordv2-server", "versions.keep");
-		this.removeBundle(item, source, keep);
+		this.removeBundle(context, item, source, keep);
 	}
 
-	public void removeBundle(Item item, Bundle source, boolean archive)
+	public void removeBundle(Context context, Item item, Bundle source, boolean archive)
 			throws SQLException, AuthorizeException, IOException
 	{
         // archive the bundle contents if desired
 		if (archive)
 		{
-			this.archiveBundle(item, source);
+			this.archiveBundle(context, item, source);
 		}
 
         // remove all the bitstreams from the bundle
-		for (Bitstream bitstream : source.getBitstreams())
+		List<BundleBitstream> bundleBitstreams = source.getBitstreams();
+		for (BundleBitstream bundleBitstream : bundleBitstreams)
 		{
-			source.removeBitstream(bitstream);
+			bundleService.removeBitstream(context, source, bundleBitstream.getBitstream());
```

Comment:
I think that we will get an "ConcurrentModificationException" here. Because the "BundleBitstream" list is persisted, so when we delete an item from a list we are iterating it will alter the list.
The best way to fix this is by using an iterator (see comment above for example)